consensus/src/detail/consensus_validator.rs

```diff
@@ -0,0 +1,90 @@
+use common::chain::block::Block;
+use common::chain::block::BlockIndex;
+use common::chain::block::ConsensusData;
+use common::chain::config::ChainConfig;
+use common::chain::ConsensusStatus;
+use common::primitives::BlockHeight;
+use common::primitives::Id;
+
+use crate::detail::pow::work::check_pow_consensus;
+use crate::BlockError;
+
+/*
+use thiserror::Error;
+#[derive(Debug, PartialEq, Eq, Error)]
+pub enum ConsensusValidationError {
+    #[error("Block consensus data is {block_consensus_data:?}, current config is {chain_consensus_status:?}")]
+    ConsensusTypeMismatch {
+        block_consensus_data: ConsensusData,
+        chain_consensus_status: ConsensusStatus,
+    },
+}
+*/
+
+/*
+impl From<ConsensusValidationError> for BlockError {
+    fn from(error: ConsensusValidationError) -> Self {
+        BlockError::ConsensusValidationError(error)
+    }
+}
+*/
+
+pub(crate) trait BlockIndexHandle {
+    fn get_block_index(
+        &self,
+        block_index: &Id<Block>,
+    ) -> blockchain_storage::Result<Option<BlockIndex>>;
+    fn get_ancestor(
+        &self,
+        block_index: &BlockIndex,
+        ancestor_height: BlockHeight,
+    ) -> Result<BlockIndex, BlockError>;
+}
+
+pub(crate) fn validate_consensus(
+    chain_config: &ChainConfig,
+    block: &Block,
+    block_index_handle: &dyn BlockIndexHandle,
+) -> Result<(), BlockError> {
```

Comment:
We also need an accessor for the utxo set. This is not a priority and we can postpone this whole thing. But feel free to do it if you have nothing else to do.

The accessor should have the functions:

- Check if a utxo is spent
- Check if a utxo is spendable

Currently this information can be retrieved from the transaction index.

Reply:
I'm resolving this as it's gonna be done for PoS at some point later.